### PR TITLE
BUG: `DataFrameGroupBy.value_counts` includes non-observed categories of non-grouping columns

### LINKED ISSUE #46357: BUG: value_counts(observed=True) inconsistent between SeriesGroupBy and DataFrameGroupBy
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [ ] I have confirmed this bug exists on the main branch of pandas.

### Reproducible Example

```python
>>> s = pd.Series(["a", "b", "c"], dtype="category").iloc[0:2]
>>> s
0    a
1    b
dtype: category
Categories (3, object): ['a', 'b', 'c']
>>> s.groupby(level=0).value_counts()
0  a    1
   b    0
   c    0
1  b    1
   a    0
   c    0
dtype: int64
>>> s.groupby(level=0, observed=True).value_counts()
0  a    1
   b    0
   c    0
1  b    1
   a    0
   c    0
dtype: int64
```

### Issue Description

The result contains `0` values for unused categories in each group with `observed=True` , unlike `DataFrameGroupBy`:
```python
>>> df = pd.DataFrame(s)
>>> df
	0
0	a
1	b
>>> df.groupby(level=0, observed=True).value_counts()
0  a    1
1  b    1
dtype: int64
```

### Expected Behavior

The behavior should be the same for `SeriesGroupBy` and `DataFrameGroupBy`. I am not sure what should be the expected behavior here. 

The `groupby` documentation specifies that `observed` parameter has effect on groupers. Considering `by` and `level` parameters are the groupers, it might make sense that in this case `GroupBy.value_counts` is not affected by the `observed` keyword and it follows the behavior of `Series.value_counts` which does keep unused categories. It would mean that `GroupBy.value_counts` (and probably `Series/DataFrame.value_counts` too) would require an additional `observed` parameter, like the suggestion in https://github.com/pandas-dev/pandas/issues/43498.

On the other hand, if we consider that the `observed` parameter makes sense in this case, `SeriesGroupBy.value_counts(observed=True)` should not return `0` values for unused categories as for `DataFrameGroupBy`.

Related: performance issue on `GroupBy.value_counts` on categoricals [#46202](https://github.com/pandas-dev/pandas/issues/46202)

### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit           : 06d230151e6f18fdb8139d09abf539867a8cd481
python           : 3.9.1.final.0
python-bits      : 64
OS               : Windows
OS-release       : 10
Version          : 10.0.19041
machine          : AMD64
processor        : Intel64 Family 6 Model 142 Stepping 11, GenuineIntel
byteorder        : little
LC_ALL           : None
LANG             : None
LOCALE           : fr_FR.cp1252

pandas           : 1.4.1
numpy            : 1.20.1
pytz             : 2021.1
dateutil         : 2.8.1
pip              : 20.3.3
setuptools       : 52.0.0.post20210125
Cython           : None
pytest           : None
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : 3.0.3
IPython          : 8.1.1
pandas_datareader: None
bs4              : None
bottleneck       : None
fastparquet      : None
fsspec           : None
gcsfs            : None
matplotlib       : None
numba            : None
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pyreadstat       : None
pyxlsb           : None
s3fs             : None
scipy            : None
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : None
xlrd             : None
xlwt             : None
zstandard        : None

</details>

### PR DESCRIPTION
**Pull request by LucasG0**
- [X] closes #46357
- [X] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [X] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [X] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

File: pandas/core/groupby/generic.py
Comment (LucasG0) [inline pandas/core/groupby/generic.py:1778]: I also checked to make this change directly in the method `_reindex_output`, so it handles the fact that groupings might have a different `observed` attribute, but it breaks 60 tests.

Comment (LucasG0) [inline pandas/core/groupby/generic.py:1778]: I also tried to make this change directly in `_reindex_output` method, so it handles the fact that groupings might have a different observed attribute, but it breaks 60 tests.

**Comment by LucasG0** [comment]
This PR has not been reviewed for 1 month. @rhshadrach as you were involved in the issue description, would you like to have a look, or should I wait/ping for the review from a specific maintainer ?

File: pandas/core/groupby/generic.py
Comment (rhshadrach) [reply pandas/core/groupby/generic.py:1778]: Makes sense that reindex_output wouldn't work here; the schema for groupby.value_counts differs from other ops.

File: pandas/tests/groupby/test_frame_value_counts.py
Comment (rhshadrach) [inline pandas/tests/groupby/test_frame_value_counts.py:329]: Why is this test changing values? In particular the order.

File: pandas/tests/groupby/test_frame_value_counts.py
Comment (rhshadrach) [inline pandas/tests/groupby/test_frame_value_counts.py:365]: We're going from a single grouper to multiple groupers here; instead, can you add a multiple groupers case so as to have both?

File: pandas/tests/groupby/test_frame_value_counts.py
Comment (rhshadrach) [inline pandas/tests/groupby/test_frame_value_counts.py:399]: There are no NaN values though?

File: pandas/tests/groupby/test_frame_value_counts.py
Comment (rhshadrach) [inline pandas/tests/groupby/test_frame_value_counts.py:404]: Add issue number as a comment on the first line of the test - `# GH#46798`

File: pandas/core/groupby/generic.py
Comment (rhshadrach) [reply pandas/core/groupby/generic.py:1778]: Does L1761 `observed=self.observed` actually do anything? I'd hazard a guess that the groupby code is just using the `observed=False` from L1752. If that is the case, can you make L1761 set observed to False as well for readability.

File: doc/source/whatsnew/v1.5.0.rst
Comment (rhshadrach) [inline doc/source/whatsnew/v1.5.0.rst:466]: ``:meth:`.DataFrameGroupBy.value_counts` `` (the extra `.`) to get the link to resolve.

**Review by rhshadrach — CHANGES_REQUESTED** [review]
Thanks for the ping @LucasG0; looks good. Some requests/questions.

File: pandas/tests/groupby/test_frame_value_counts.py
Comment (LucasG0) [reply pandas/tests/groupby/test_frame_value_counts.py:329]: I modified this test so it uses multiple groupers (`country` and `education` instead of `country` only), thus the second index level is the `education` column and is sorted by the groupby `sort=True` default behavior.

I will remove the `as_index` change.

File: pandas/core/groupby/generic.py
Comment (LucasG0) [reply pandas/core/groupby/generic.py:1778]: `observed=self.observed` is needed for categorical grouping columns. `observed=False` is used L1752 so every category of any categorical non-grouping column is included in the result.

File: pandas/tests/groupby/test_frame_value_counts.py
Comment (LucasG0) [reply pandas/tests/groupby/test_frame_value_counts.py:365]: Yes I'll restore the original single grouper test

File: pandas/tests/groupby/test_frame_value_counts.py
Comment (LucasG0) [reply pandas/tests/groupby/test_frame_value_counts.py:399]: No, the filling value is 0 - if a key is not present, it appears 0 time.

File: pandas/core/groupby/generic.py
Comment (rhshadrach) [reply pandas/core/groupby/generic.py:1778]: Ah, thanks - I forgot the above groupers on L1752 are only the non-grouping columns.

File: pandas/tests/groupby/test_frame_value_counts.py
Comment (rhshadrach) [reply pandas/tests/groupby/test_frame_value_counts.py:399]: Right - that the filling value is 0 makes sense. The comment about NaN values still does not to me though. What NaNs is it referring to?

File: pandas/core/groupby/generic.py
Comment (jreback) [inline pandas/core/groupby/generic.py:1777]: what is the fill_value hitting? is there a test that needs this?

**Review by jreback — COMMENTED** [review]
minor comment.

File: pandas/tests/groupby/test_frame_value_counts.py
Comment (LucasG0) [reply pandas/tests/groupby/test_frame_value_counts.py:399]: Oh sorry this comment is obsolete, I will remove it.

File: pandas/tests/groupby/test_frame_value_counts.py
Comment (LucasG0) [reply pandas/tests/groupby/test_frame_value_counts.py:516]: @rhshadrach I added tests for the single grouper case. They actually tests when both groupers and non-groupers are categorical, as this path was not tested in the multiple groupers test.

File: pandas/core/groupby/generic.py
Comment (LucasG0) [inline pandas/core/groupby/generic.py:1805]: It fixes an other bug related to `normalize` incorrectly producing NaN on non-observed categories:
```
>>> c1 = pd.Categorical(["a", "b"], categories=["a", "b", "c"])
>>> c2 = range(2)
>>> df = pd.DataFrame({"c1": c1, "c1": c2})
>>> df.groupby("c1").value_counts(normalize=True)
c1  c2
a   0     1.0
    1     0.0
b   1     1.0
    0     0.0
c   0     NaN    # <--- 0 is expected
    1     NaN    # <--- 0 is expected
dtype: float64
```

File: pandas/core/groupby/generic.py
Comment (LucasG0) [reply pandas/core/groupby/generic.py:1787]: The purpose of `fill_value=0` is to fill non-observed categories count with 0 as they are not present in the `result_series`, and this behavior is tested in the added tests. Does it answer your question ?

File: pandas/core/groupby/generic.py
Comment (jreback) [reply pandas/core/groupby/generic.py:1805]: I guess these are by definition NaN here as we reindexed. this seems quite unfortunate as .value_counts() should be integer valued. can we fill on the reindex iteself? (which will keep the int dtype)

File: pandas/tests/groupby/test_frame_value_counts.py
Comment (jreback) [inline pandas/tests/groupby/test_frame_value_counts.py:311]: name this `assert_categorical_single_grouper` or similar, this is *not* a test rather an asserter

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/groupby/generic.py
Comment (rhshadrach) [reply pandas/core/groupby/generic.py:1805]: When normalize is True (this section of code), the values are percentages in [0.0, 0.1]. The nans don't come from a reindex, rather here they are coming from where `indexed_group_size = result_series.groupby(...).transform('sum')` is 0.

File: pandas/tests/groupby/test_frame_value_counts.py
Comment (LucasG0) [reply pandas/tests/groupby/test_frame_value_counts.py]: Done

**Comment by LucasG0** [comment]
@jreback @rhshadrach do you have any other comments to add on this PR or should we consider merging it (after rebase) ?

File: pandas/tests/groupby/test_frame_value_counts.py
Comment (rhshadrach) [reply pandas/tests/groupby/test_frame_value_counts.py:479]: It looks like you've added the single grouper test back but with changes - this makes the diff very hard to follow. Would it be an issue to restore the original test fully?

**Review by rhshadrach — CHANGES_REQUESTED** [review]
Overall looks good, I'm not sure why the one test is still changing though.

File: doc/source/whatsnew/v1.5.0.rst
Comment (jreback) [inline doc/source/whatsnew/v1.5.0.rst:637]: ok since this is changing non-trivially, can you make a sub-section (ok here is fine, where you show the prior and the fixed behavior)

**Review by jreback — CHANGES_REQUESTED** [review]

File: doc/source/whatsnew/v1.5.0.rst
Comment (LucasG0) [reply doc/source/whatsnew/v1.5.0.rst:758]: I added a sub-section in notable fixes

File: pandas/tests/groupby/test_frame_value_counts.py
Comment (LucasG0) [reply pandas/tests/groupby/test_frame_value_counts.py:539]: It just adds some slight redundancy with other single grouper tests. I restored the original test, now called `test_categorical_single_grouper_with_only_observed_categories`, with a small change to `observed=True` case regarding this fix

**Comment by LucasG0** [comment]
Unsuccessful job seems to be related to the flakyness of https://github.com/pandas-dev/pandas/pull/45651

File: pandas/tests/groupby/test_frame_value_counts.py
Comment (rhshadrach) [reply pandas/tests/groupby/test_frame_value_counts.py:539]: @LucasG0 - there is too much changing here. Correct me if I'm wrong, but it appears the original test is not restored - it seems to be heavily modified.

Comment (LucasG0) [reply pandas/tests/groupby/test_frame_value_counts.py:539]: The original test on the main branch is called `test_categorical`, to be more explicit and to avoid confusion with other added tests, it is now called `test_categorical_single_grouper_with_only_observed_categories` ([link](https://github.com/pandas-dev/pandas/pull/46798/files#diff-94eb7e512970f337748151525e12830e7ca10c5e8acf4db15a00f56a9b2e6814R323)) and I added it in [this commit](https://github.com/pandas-dev/pandas/pull/46798/commits/fc8ea62187c89b8fd92aba071e54e378089edab4).

I slightly modified the original test in this PR to fix the expected behavior with `observed=True`. In the original test, the result would differ depending on `observed` while it should not - non grouping columns are not affected by `observed`, and grouping columns do not have non-observed values here, thus `observed` has no effect.

**Comment by simonjayhawkins** [comment]
@jreback @rhshadrach `DataFrameGroupBy.value_counts` was added in pandas-1.4.0

Firstly. do we need the extensive release note. These are normally reserved for bug fixes that have a significant change in behavior that might be considered a breaking change by users.

Also because `DataFrameGroupBy.value_counts` is a new feature we could consider backporting to 1.4.x.

File: doc/source/whatsnew/v1.5.0.rst
Comment (simonjayhawkins) [inline doc/source/whatsnew/v1.5.0.rst:267]: if keeping this, need `.. _whatsnew_150.notable_bug_fixes` label

**Comment by jreback** [comment]
no backport in this
and i think does need a big note as this is a significant change that's not obvious

File: doc/source/whatsnew/v1.5.0.rst
Comment (rhshadrach) [reply doc/source/whatsnew/v1.5.0.rst:758]: Sub-section looks good; need to remove this line.

**Review by rhshadrach — CHANGES_REQUESTED** [review]
Needs two fixes from the whatsnew (some from me, one from Simon); otherwise looks good.

**Comment by LucasG0** [comment]
@rhshadrach I fixed the whatsnew

**Comment by LucasG0** [comment]
Job failure is due to the unavailability of Github Actions

**Review by rhshadrach — APPROVED** [review]
lgtm

**Comment by rhshadrach** [comment]
Apologies @LucasG0 - this fell off my radar. Can you resolve conflicts.

@simonjayhawkins - are you good here?

**Comment by LucasG0** [comment]
No worries @rhshadrach, conflicts resolved.

**Comment by rhshadrach** [comment]
@simonjayhawkins - friendly ping

**Comment by simonjayhawkins** [comment]
> @simonjayhawkins - friendly ping

My comments were related to whether this should have been a targeted backport or not.

But I'm not going to block.

**Comment by rhshadrach** [comment]
@jreback - friendly ping

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @LucasG0 this is very nice! and thanks @rhshadrach for all the work on this as well.

**Pull request merged**
